File: boat-sample/src/test/java/sample/java/xyz/srclab/core/proxy/ProxySample.java
package sample.java.xyz.srclab.core.proxy;

import org.jetbrains.annotations.NotNull;
import org.testng.annotations.Test;
import xyz.srclab.common.proxy.ClassProxy;
import xyz.srclab.common.proxy.ProxyMethod;
import xyz.srclab.common.proxy.SourceInvoker;
import xyz.srclab.common.test.TestLogger;

import java.lang.reflect.Method;
import java.util.Arrays;

public class ProxySample {

    private static final TestLogger logger = TestLogger.DEFAULT;

    @Test
    public void testProxy() {
        ClassProxy<Object> proxyClass = ClassProxy.newProxyClass(
            Object.class,
            Arrays.asList(
                new ProxyMethod<Object>() {

                    @Override
                    public boolean isProxy(@NotNull Method method) {
                        return method.getName().equals("toString")
                            && Arrays.equals(method.getParameterTypes(), new Class[0]);
                    }

                    @Override
                    public Object invoke(
                        @NotNull Object proxied,
                        @NotNull Method proxiedMethod,
                        @NotNull SourceInvoker superInvoke,
                        Object @NotNull [] args
                    ) {
                        return "Proxy[super: " + superInvoke.invoke(args) + "]";
                    }
                }
            )
        );
        String s = proxyClass.newInstance().toString();
        //Proxy[super: net.sf.cglib.empty.Object$$EnhancerByCGLIB$$4926690c@256f38d9]
        logger.log("s: {}", s);
    }
}

File: boat-sample/src/test/kotlin/sample/kotlin/xyz/srclab/core/proxy/ProxySample.kt
package sample.kotlin.xyz.srclab.core.proxy

import org.testng.annotations.Test
import xyz.srclab.common.proxy.ClassProxy
import xyz.srclab.common.proxy.ProxyMethod
import java.lang.reflect.Method
import java.util.*

class ProxySample {

    @Test
    fun testProxy() {
        val proxyClass = ClassProxy.newProxyClass(
            Any::class.java,
            listOf(
                object : ProxyMethod<Any> {

                    override fun isProxy(method: Method): Boolean {
                        return method.name == "toString" && Arrays.equals(
                            method.parameterTypes,
                            arrayOfNulls<Class<*>>(0)
                        )
                    }

                    override fun invoke(
                        proxied: Any,
                        proxiedMethod: Method,
                        sourceInvoke: SourceInvoke,
                        args: Array<out Any?>
                    ): Any {
                        return "Proxy[super: " + sourceInvoke.invoke(*args) + "]"
                    }
                }
            )
        )
        val s = proxyClass.newInstance().toString()
        //Proxy[super: net.sf.cglib.empty.Object$$EnhancerByCGLIB$$4926690c@256f38d9]
        logger.log("s: {}", s)
    }

    companion object {
        private val logger = TestLogger.DEFAULT
    }
}